types/module.go

```diff
@@ -0,0 +1,233 @@
+
+Independent module functions are separated to allow for the construction of the
+basic application structures required early on in the application definition
+and used to enable the definition of full module functionality later in the
+process. This separation is necessary, however we still want to allow for a
+high level pattern for modules to follow - for instance, such that we don't
+have to manually register all of the codecs for all the modules. This basic
+procedure as well as other basic patterns are handled through the use of
+ModuleBasicManager.
+*/
+package types
+
+import (
+	"encoding/json"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/spf13/cobra"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// ModuleClient helps modules provide a standard interface for exporting client functionality
+type ModuleClient interface {
+	GetQueryCmd() *cobra.Command
+	GetTxCmd() *cobra.Command
+}
+
+//__________________________________________________________________________________________
+// AppModule is the standard form for basic non-dependant elements of an application module.
+type AppModuleBasic interface {
+	Name() string
+	RegisterCodec(*codec.Codec)
+	DefaultGenesis() json.RawMessage
+	ValidateGenesis(json.RawMessage) error
+}
+
+// collections of AppModuleBasic
+type ModuleBasicManager []AppModuleBasic
+
+func NewModuleBasicManager(modules ...AppModuleBasic) ModuleBasicManager {
+	return modules
+}
+
+// RegisterCodecs registers all module codecs
+func (mbm ModuleBasicManager) RegisterCodec(cdc *codec.Codec) {
+	for _, mb := range mbm {
+		mb.RegisterCodec(cdc)
+	}
+}
+
+// Provided default genesis information for all modules
+func (mbm ModuleBasicManager) DefaultGenesis() map[string]json.RawMessage {
+	genesis := make(map[string]json.RawMessage)
+	for _, mb := range mbm {
+		genesis[mb.Name()] = mb.DefaultGenesis()
+	}
+	return genesis
+}
+
+// Provided default genesis information for all modules
+func (mbm ModuleBasicManager) ValidateGenesis(genesis map[string]json.RawMessage) error {
+	for _, mb := range mbm {
+		if err := mb.ValidateGenesis(genesis[mb.Name()]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+//_________________________________________________________
+// AppModule is the standard form for an application module
+type AppModule interface {
+	AppModuleBasic
+
+	// registers
+	RegisterInvariants(InvariantRouter)
+
+	// routes
+	Route() string
+	NewHandler() Handler
+	QuerierRoute() string
+	NewQuerierHandler() Querier
+
+	// genesis
+	InitGenesis(Context, json.RawMessage) []abci.ValidatorUpdate
+	ExportGenesis(Context) json.RawMessage
+
+	BeginBlock(Context, abci.RequestBeginBlock) Tags
+	EndBlock(Context, abci.RequestEndBlock) ([]abci.ValidatorUpdate, Tags)
+}
+
+// module manager provides the high level utility for managing and executing
+// operations for a group of modules
+type ModuleManager struct {
+	Modules            map[string]AppModule
+	OrderInitGenesis   []string
```

Comment:
I am not a huge fan of this pattern, for two reasons:
- It's weakly typed - easy to misspell a name or forget to include one
- It's less intuitive than just calling the functions explicitly in order (imo)

I would vote to keep explicit `InitGenesis`, `ExportGenesis`, `BeginBlocker`, and `EndBlocker` functions in the application (and otherwise retain the module abstractions in this file, which I generally like). I think explicit functions in this case are going to be more obvious than this method, even though this method is a wee bit more abstract.

Open to other opinions though.

Reply:
I actually do like the ordering semantics there. Think its pretty clear and very user friendly.

Reply:
Weakly held opinion here, if the quorum is against fine with this.

Reply:
nope sounds good, I think there is a nice way to accomplish this while still keeping these functions are part of the module manager, such as calling those functions explicitly _from_ the module manager. I'll give it a whirl and then present the alternative here.

Reply:
olright - I gave it a try, and I've concluded that I think we should keep the existing structure proposed in the PR (aka use the module manager), it's becoming quite verbose in gaia with all the tag appending etc. etc. etc. you can see for yourself in the branch `rigel/genesis-generalization3-alt`

> It's weakly typed - easy to misspell a name or forget to include one

meh in gaia we're referencing all the modules by their name via the modules aka `staking.Name` this is not an issue this way. Also I think it's _more_ likely that you will forget to add something in if we call them all explicitly because these definitions are scattered all over the place (for export actually in a different file) as opposed to tightly and neatly in the same place where the module manager is defined